### ISSUE TITLE
PINVOKE Error With Most Recent ExportToOpenStudio

### ISSUE THREAD

**Issue by chriswmackey** [ladybug-tools/honeybee-legacy#214]
@chiensiTB , Do you know why I get this error with the OpenStudio Component that you just committed?

![image](https://cloud.githubusercontent.com/assets/5567574/4877830/d4dc72a4-62f8-11e4-98ae-75127a1d3ac4.png)

**Comment by macumber** [comment]
@mostaphaRoudsari @chiensiTB I tried looking into this today.  I downloaded the newest version of Ladybug/Honeybee from food4rhino, removed my old Ladybug/Honeybee, and updated to the new versions.  I removed my old C:\ladybug\OpenStudio, added a new C:\Users\dmacumbe\AppData\Roaming\Ladybug\OpenStudio, and copied the linked OpenStudio dlls over there.  When I try to run the testImportingOpenStudioLibs.gh script I get this output:

Runtime error (IOException): file does not exist: C:\Users\dmacumbe\AppData\Roaming\Ladybug\OpenStudio\openStudio.dll
Traceback:
  line 12, in script

I corrected the case of OpenStudio.dll but still get the same error.  The file does exist on my computer.  I suspect some other problem is happening, see this post http://ironpython.codeplex.com/workitem/31749.

Is this the latest error you guys are seeing?

**Comment by mostaphaRoudsari** [comment]
@macumber thank you for looking into this. No we don't have the issue of finding the file. Did you unblock your .dlls? (Right click on the file > unblock?)

try this [sample file](https://www.dropbox.com/s/p6ze63n3k3ctmqt/testImportingOpenStudioLibs.gh?dl=0). It only tries to import the library. Thanks again and let me know if you want to set up a googleHangout and try to figure it out together. This will be a huge help to us to go forward.

**Comment by macumber** [comment]
Hey Mostapha, I had problems with the OpenStudio package that OpenStudioExample.gh pointed me to, https://app.box.com/s/y2sx16k98g1lfd3r47zi.

Instead, I downloaded the 1.6.0 64bit OpenStudio installer from https://www.openstudio.net/downloads, ran the installer, then copied all the files from 'C:\Program Files\OpenStudio 1.6.0\CSharp\openstudio' to 'C:\Users\dmacumbe\AppData\Roaming\Ladybug\OpenStudio'.

After that testImportingOpenStudioLibs.gh works fine, and OpenStudioExample.gh works but I get this message (I don't know how to change units):

Currently the OpenStudio component only works in meters. Change the units to Meters and try again!

If you guys use the OpenStudio dlls from our installer do you still run into pInvoke problems?

**Comment by macumber** [comment]
Ok, figured out how to change units and ran the simulation :-)

**Comment by mostaphaRoudsari** [comment]
@macumber you imported the latest version of OpenStudio with no problem! That's great. Can you share the libraries with us so we can also try it on our systems?

The most recent version of Honeybee on github actually works with both 8.1 and 8.2

**Comment by macumber** [comment]
@mostaphaRoudsari like I said, just download the installer from https://www.openstudio.net/downloads, run the installer, then copy all the files from 'C:\Program Files\OpenStudio 1.6.0\CSharp\openstudio' to 'C:\Users\#{your_user}\AppData\Roaming\Ladybug\OpenStudio'.

**Comment by mostaphaRoudsari** [comment]
@macumber as far as I can see the site is under maintenance

![image](https://cloud.githubusercontent.com/assets/2915573/6365100/5446c906-bc7f-11e4-84ba-4276429bd323.png)

but I have also a question. Are you shipping 64bits library also with installation? I remember it used to be 32bits which wasn't working for us.

**Comment by macumber** [comment]
Weird, the website was working for me.  You can also get downloads on our github page, here is a link to the 1.6.0 release, https://github.com/NREL/OpenStudio/releases/tag/v1.6.0

Direct link to 64 bit installer, https://github.com/NREL/OpenStudio/releases/download/v1.6.0/OpenStudio-1.6.0.1229eac298-Win64.exe

Our default installer on Windows is now 64 bit since SketchUp 2015 is now 64 bit

**Comment by mostaphaRoudsari** [comment]
@macumber Thanks. Downloading right now! I'll let you know if it works on my system.

@chiensiTB, @chriswmackey and @jacoposartore can you guys also test on your systems and let us know the results?

**Comment by mostaphaRoudsari** [comment]
@macumber  :( - Still the same error on my system.

![image](https://cloud.githubusercontent.com/assets/2915573/6365994/47876e4c-bc8d-11e4-9f30-000c7fdc7550.png)

**Comment by macumber** [comment]
Hmm, bummer.  I am uploading a copy of my debug build dlls to this link (will take some time to upload), https://www.dropbox.com/s/oczwi4yylxixcvg/OpenStudio_1_6_3.zip?dl=0

Maybe those will give you some more information.  If not, I think the thing to do would be to have you build your own 64 bit OpenStudio bindings and try those out, that will give you better debug information.

The only other idea I have is instead of appending the OpenStudio dir to the end of your path, you might try adding at the front of the path.

**Comment by macumber** [comment]
Ok this is definitely a path issue, I was just able to reproduce by doing this at the command line:

C:\Program Files\Rhinoceros 5 (64-bit)\System>set PATH=""
C:\Program Files\Rhinoceros 5 (64-bit)\System>Rhino.exe

**Comment by mostaphaRoudsari** [comment]
Are you talking about sys.path or are you talking about my windows system path?
I added the path as the first item to sys.path and it is still the same

![image](https://cloud.githubusercontent.com/assets/2915573/6366311/c4a9e9f0-bc91-11e4-9e3c-05d70c8e8217.png)

**Comment by macumber** [comment]
Have you tried adding this line:

os.environ['PATH'] = os.environ['PATH'] + ";" + openStudioLibFolder

Looks like sys.path is just the Python path, does not actually affect the system path which is where dependent libraries are loaded from

**Comment by mostaphaRoudsari** [comment]
Just tried and it doesn't help. the same error. There should be something else in your Windows path that I don't have in mine. I hope the debug version tells us something about that...

**Comment by jacoposartore** [comment]
hi @mostaphaRoudsari I just have tried on my system what @macumber did (copying the dlls within C:\Program Files\OpenStudio 1.6.0\CSharp\openstudio into c:\ladybug\openstudio) and now both the testImportingOpenStudioLibs.gh and the OpenStudioExample.gh are working fine.

![testimportingopenstudiolibs](https://cloud.githubusercontent.com/assets/8501837/6367359/68d5ea5c-bcd2-11e4-8e3a-b741a67d8efc.png)

In particular, as regards the last file, the Export to OpenStudio component returns an orange error about not having the latest OpenStudio libraries but the .osm file is eventually generated as you can see from the screenshot.

![openstudioexample](https://cloud.githubusercontent.com/assets/8501837/6367403/0da89c64-bcd3-11e4-91ff-bece5ea4125f.png)


![openstudioopenexample](https://cloud.githubusercontent.com/assets/8501837/6367404/12cb20e0-bcd3-11e4-8fb2-228cf805f2bf.jpg)

**Comment by macumber** [comment]
Hmm @mostaphaRoudsari, in my testing I reproduced your error by completely wiping out my PATH environment variable before starting Rhino.  Then I fixed the error by adding the openStudioLibFolder to the PATH environment variable (Rhino did add some Rhino paths to this variable when it started) in the script before loading the assembly file, here is the code that did that:

os.environ['PATH'] = os.environ['PATH'] + ";" + openStudioLibFolder

I guess that you could test by adding openStudioLibFolder to your regular windows PATH, but I think my fix will be better for most of your users as they won't have to modify their PATH permanently.  Could you try again to see if this works for you?

**Comment by mostaphaRoudsari** [comment]
@macumber I tested on my work machine and it works fine. Still have the issue on my laptop. Also used the new debug libraries and it generates similar error report. Nothing really helpful.

**Comment by macumber** [comment]
Thanks everyone for testing, @mostaphaRoudsari I missed some DLLs in my first zip of debug libraries, can you re-download this one and try again on your laptop?

https://www.dropbox.com/s/oczwi4yylxixcvg/OpenStudio_1_6_3.zip?dl=0

Also, it is possible there is something in your path that is causing an issue (alternate versions of Qt or OpenSSL would be my guess).  Can you open a command prompt in your Rhino folder and type these commands:

C:\Program Files\Rhinoceros 5 (64-bit)\System>set PATH=""
C:\Program Files\Rhinoceros 5 (64-bit)\System>set PYTHONPATH=""
C:\Program Files\Rhinoceros 5 (64-bit)\System>Rhino.exe

That will launch Rhino from an empty environment, then when you run the code you will be starting from a clean start.  It's ugly but in OpenStudio we get around these issues by always adding our stuff to the front of the path when our applications load.

**Comment by mostaphaRoudsari** [comment]
:|. I cleaned it up inside the code and also tried your solution and in both cases it doesn't work. as you can see I only have c:\openstudio\openstudio in both Windows path and python's sys.path

![image](https://cloud.githubusercontent.com/assets/2915573/6379502/da64766a-bd00-11e4-8996-9b477b7939a8.png)

**Comment by mostaphaRoudsari** [comment]
@macumber I'm uninstalling OpenSSL and Qt! Let's see if that helps.

**Comment by macumber** [comment]
Can you try this code:

```python
import scriptcontext as sc
import os

if sc.sticky.has_key('honeybee_release'):
    
    openStudioLibFolder = os.path.join(sc.sticky["Honeybee_DefaultFolder"], "OpenStudio")
    
    if os.path.isdir(openStudioLibFolder) and os.path.isfile(os.path.join(openStudioLibFolder, "OpenStudio.dll")):
        
        print os.environ['PATH']
        os.environ['PATH'] = openStudioLibFolder + ";" + os.environ['PATH']
        print os.environ['PATH']
        
        import sys
        
        if openStudioLibFolder not in sys.path:
            sys.path.append(openStudioLibFolder)
        
        import clr
        clr.AddReferenceToFileAndPath(openStudioLibFolder+"\\OpenStudio.dll")
        
        import OpenStudio as ops

        print ops.Model()
    else:
        print "Can't find OpenStudio library in " + openStudioLibFolder
```

**Comment by mostaphaRoudsari** [comment]
No luck! I also removed Qt and uninstalled OpenSSL

![image](https://cloud.githubusercontent.com/assets/2915573/6379848/4061394c-bd03-11e4-8c1c-e577dfc6904e.png)

**Comment by macumber** [comment]
I see a bunch of other stuff in your path, are you using the command prompt to clear PATH and PYTHON path (will only affect that command prompt session) and start Rhino from a clean environment?

**Comment by mostaphaRoudsari** [comment]
I copied my second try. Here is the first one with clean path:

![image](https://cloud.githubusercontent.com/assets/2915573/6380133/6567f7ce-bd05-11e4-82d2-a85cae645e39.png)

**Comment by macumber** [comment]
@mostaphaRoudsari this is slightly unrelated, but going forward I would suggest that you try to use the OpenStudio C# bindings directly from an OpenStudio installation.  That way your directions to your users are "Install OpenStudio version XXX" then point Ladybug to it somehow.  This will have the added benefit of allowing you to find the version of E+ and Ruby that get installed with OpenStudio, which will also allow you run measures at some point in the future.

**Comment by macumber** [comment]
@mostaphaRoudsari this is really frustrating, sorry about that.  Just so we know we are all on the same page, can you install OS 1.6.0, launch Rhino from a clean environment, and try this code:

```python
import scriptcontext as sc
import os

if sc.sticky.has_key('honeybee_release'):
    
    #openStudioLibFolder = os.path.join(sc.sticky["Honeybee_DefaultFolder"], "OpenStudio")
    openStudioLibFolder = "C:\\Program Files\\OpenStudio 1.6.0\\CSharp\\openstudio"
    
    if os.path.isdir(openStudioLibFolder) and os.path.isfile(os.path.join(openStudioLibFolder, "OpenStudio.dll")):
        
        print os.environ['PATH']
        os.environ['PATH'] = openStudioLibFolder + ";" + os.environ['PATH']
        print os.environ['PATH']
        
        import sys
        
        if openStudioLibFolder not in sys.path:
            sys.path.append(openStudioLibFolder)
        
        import clr
        clr.AddReferenceToFileAndPath(openStudioLibFolder+"\\OpenStudio.dll")
        
        import OpenStudio as ops

        print ops.Model()
    else:
        print "Can't find OpenStudio library in " + openStudioLibFolder
```

**Comment by mostaphaRoudsari** [comment]
Still the same! Maybe I should do a comparison with my work machine?

![image](https://cloud.githubusercontent.com/assets/2915573/6380610/d56a38e0-bd08-11e4-9600-be05395e6625.png)

**Comment by mostaphaRoudsari** [comment]
```
@mostaphaRoudsari this is slightly unrelated, but going forward I would suggest that you try to use the OpenStudio C# bindings directly from an OpenStudio installation. That way your directions to your users are "Install OpenStudio version XXX" then point Ladybug to it somehow. This will have the added benefit of allowing you to find the version of E+ and Ruby that get installed with OpenStudio, which will also allow you run measures at some point in the future.
```

Totally! We had it this way since the OpenStudio's default release was not 64bits at the time. Thanks.

**Comment by mostaphaRoudsari** [comment]
Just for the record the first time that we wanted to get the libraries loaded we emailed 70 times:

![image](https://cloud.githubusercontent.com/assets/2915573/6380779/1f3b1416-bd0a-11e4-835a-54dcd2c83128.png)

https://docs.google.com/presentation/d/1wexas-XWMWaQdEVpOmPP8oVTR3D2xRhNtLAw7M-UzsA/edit#slide=id.ge3cf2cba_6133

Too soon to give up! ;)

**Comment by macumber** [comment]
@mostaphaRoudsari you wrote half of those emails so thank you for not giving up on us!

**Comment by mostaphaRoudsari** [comment]
:+1: Never! I'm sure we'll get this one fixed too at some point.

**Comment by macumber** [comment]
@mostaphaRoudsari @chiensiTB are you guys still having this issue?

**Comment by mostaphaRoudsari** [comment]
@macumber I still have this issue on my laptop! :|

**Comment by mostaphaRoudsari** [comment]
This is [potentially] a good news! I downloaded OpenStudio1.8.0 today and tested the libraries and it worked with no issues on my machine. @chriswmackey and @chiensiTB can you guys also try and let me know. This can be a great news for all of us. The development for OpenStudio has been on hold for awhile because of this issue.

**Comment by chriswmackey** [comment]
Software gods be praised! It works!

I am going to draw up a new github issue for all of the backlogged features that we need to work into the OpenStudio exporter.  Finally, it seems a dark age of development has passed!

-Chris

**Comment by chiensiTB** [comment]
Oh shit. Does this mean I don't have to be the ugly black sheep of the family anymore?

**Comment by mostaphaRoudsari** [comment]
I guess so! Also now the defaults for HVAC is supposedly way much better than what it is used to be, and apparently there is a going to be a json for the ASHRAE HVAC types > https://github.com/NREL/OpenStudio-Prototype-Buildings#resourceshvacsizing

**Comment by macumber** [comment]
Well this is great news!  I'm curious what might have changed between 1.7.0 and 1.8.0, I can't think of anything (dependencies, etc) that might have fixed this.  @axelstudios can you think of anything that changed in the OpenStudio installer?

**Comment by axelstudios** [comment]
I was wondering that myself - All I can guess is that maybe it was resolved with Swig 3.0.5 (3.0.6 just came out btw)

**Comment by chriswmackey** [comment]
It's time to take care of some major backlogged development!!
https://github.com/mostaphaRoudsari/Honeybee/issues/359
-Chris

**Comment by jacoposartore** [comment]
I just wanted to add that I had to update OpenStudio to version 1.8.1 (and so the openstudio libraries) in order to get the example file working fine (https://www.dropbox.com/s/pwpsravhwk0vlal/OpenStudioExample.gh?dl=0). Did you have this kind of error? However both the versions imported correctly the libraries (https://www.dropbox.com/s/p6ze63n3k3ctmqt/testImportingOpenStudioLibs.gh?dl=0).

**Comment by mostaphaRoudsari** [comment]
Thanks @jacoposartore. Strange! It works also with 1.8.0 on my system. In any case as far as it works with the newest version I will be a happy guy! :tada: :balloon: 

@chiensiTB can you also try it and let us know so we can close this long-time open unsolvable issue!

**Comment by jacoposartore** [comment]
@mostaphaRoudsari I did some tests and I had to revert to OpenStudio 1.7.0: I figured out that when my definition is using the 1.8.1 libraries is generating a file that in OpenStudio is missing some fields (like the Water Loops) and the simulation fails. With the libraries from version 1.7.0  it has all the expected HVAC systems and the simulation runs fine.

**Comment by chiensiTB** [comment]
Ok.  Looks like we are back on!  Everything is running fine.  I did have to update all the references to 8-1-0 to 8-3-0 in the OpenStudio component.  I don't know if anyone else had this issue.  I guess it just depends on what version of E+ you have running.  But this is great news!  New life....

**Comment by jacoposartore** [comment]
@mostaphaRoudsari I did another test today and it's now working with 1.8.0 libraries. Strange! Maybe I forgot to copy some dll.

**Comment by mostaphaRoudsari** [comment]
Perfect! Thank you all. I'm closing this with so much pleasure.

@jacoposartore back to your comment about systems such a Water Loops if there is any case that doesn't work please report them as an issue. They might be a bug in OpenStudio or a change in the API so we can update accordingly.

**Comment by antonszilasi** [comment]
@chriswmackey @mostaphaRoudsari @macumber 

Gentlemen how do I actually solve this issue? I am experiencing it with Open Studio version 1.9 at the moment. Where can I download an earlier release of Open Studio? My understanding from this discussion is that using version 1.8 should fix the issue.

**Comment by mostaphaRoudsari** [comment]
@antonszilasi did you try what I suggested on Grasshopper group? You should not see the issue if you're using 1.9+ OpenStudio libraries. You can download older versions from github. Check release tags here: https://github.com/NREL/OpenStudio/tags

**Comment by chriswmackey** [comment]
@antonszilasi ,
Did you conclude that the error was an idiosyncrasy on your system or has the PINVOKE Error returned?  Please let me know and, if it is the former, I will update to 1.9 now.

**Comment by antonszilasi** [comment]
@chriswmackey  when I updated to OpenStudio 1.9.5 it went away. I had to manipulate the code within Export to Open Studio to get 1.9.5 to work however.

**Comment by chriswmackey** [comment]
Great gurus of energy modelling software,

It is with a grave heart that I report that this issue has not been fully stamped-out.  It dangles to life by a thread but we have one confirmed user (in a total of maybe a thousand that have used the OpenStudio component) that still faces this issue in OpenStudio 1.12.0:
http://www.grasshopper3d.com/group/ladybug/forum/topics/cannot-export-to-openstudio?commentId=2985220%3AComment%3A1586957&xg_source=msg_com_gr_forum

The error is exactly the same as it has been reported here with a failure happening when the model object is initiated.  We were hoping to completely phase out our IDF-writing component in order to direct all users to OpenStudio but it seems we might not be able to do this unless we find a solution here.

@macumber and @axelstudios , given that the SWIG update seems to have fixed this issue in the past, are there any plans on the horizon for another SWIG update in the coming version of OpenStudio?

Thank you,
-Chris

**Comment by Worrily** [comment]
Omg, I`ve been suffering from this issue as well. Is there any possible solution to this issue? OpenStudio has updated to 2.6.0 in 2018 now, the problem still exists. And I`ve read all things you`ve talked about in 2016, and have tried things you said, nothing worked. I`m so depressed.

**Comment by CongzhengZOU** [comment]
Hi, all, sorry to disturb all of you. Now we are in November 2020, I have the same error.
I have Rhino 6, Ladybug tools, Openstudio 3.0.1, Radiance 5.1.0 installed.
And I am just jusing the example file " [Multizone_Energy_Balance_Chart](http://hydrashare.github.io/hydra/viewer?owner=chriswmackey&fork=hydra_2&id=Multizone_Energy_Balance_Chart&slide=0&scale=5.278031643091579&offset=-2828.1801646555423,-854.9622347123345) "in Hydra
![Error-Multizone_Energy_Balance_Chart](https://user-images.githubusercontent.com/70321029/98557201-2f641080-22a4-11eb-9f89-99dc382ea269.png)

**Comment by mostaphaRoudsari** [comment]
@CongzhengZOU, see the forum. This is a very old issue and has been resolved.

https://discourse.ladybug.tools
